### DISCUSSION TITLE
Exhaustability for switch expressions on enums should be less strict

### DISCUSSION THREAD

**Discussion by YairHalberstadt** [dotnet/csharplang#2671] [Language Ideas]
Currently switch expressions emit a warning if they are not exhaustive.

The following will warn:

```csharp
enum Reply
{
    Yes,
    No,
}
...
Reply reply = GetReply();
reply switch
{
     Reply.Yes => true,
     Reply.No => false,
};
```

The reason for this is that this enum can be any value of type int, not just its declared values.

This is only a warning, so the compiler can still compile the code. This is what the compiled code looks like:

```csharp
reply switch
{
     Reply.Yes => true,
     Reply.No => false,
     _ => throw new InvalidOperationException(),
};
```

It throws in the unexpected case.

I think it can safely be assumed that the aim of this warning is to reduce bugs. There are two choices on how strict to be in exhaustiveness checking:

a)    **Pragmatic:** A switch expression is considered exhaustive if all declared values for the enum are matched. This will lead to exceptions being thrown at runtime if the enum turns out to be some other value.

b)     **Strict:** A switch expression is considered exhaustive if all possible values for the enum are matched. In practice the means adding a `_ =>` arm to the switch, since for anything other than a `byte` based enum this is impossible. This will lead to developers not being warned if a new value is added to the enum which they do not account for.

We need some data to help drive this decision. I went through every single example in Roslyn where a switch expression on an enum handles every single declared value of the enum. They all currently have a `_ =>` arm to remove the warning.

Here are all the methods where they are:

```
NullableAnnotationExtensions.MergeNullableAnnotations
NullableAnnotationExtensions.ToPublicAnnotation
NullableAnnotationExtensions.ToInternalAnnotation
NullableFlowStateExtensions.ToPublicFlowState
NullableFlowStateExtensions.ToInternalFlowState
SeparatedSyntaxListCodeActionComputer.GetNestedCodeActionTitle
CSharpTriviaResult.AnnotationResolver
CSharpTriviaResult.TriviaResolver
CSharpUseRangeOperatorCodeFixProvider.CreateRangeExpression
ExecutionConditionUtil.Architecture
ProjectExternalErrorReporter.ReportError2
VSTelemetryLogger.CreateAndStartScope
VSTelemetryLogger.LogBlockEnd
BaseDiagnosticItem.BrowseObject.MapDiagnosticSeverityToText
BaseDiagnosticItem.BrowseObject.MapReportDiagnosticToText
CSharpProjectShim.OptionsProcessor.SetOutputFileType
FindAllDeclarationsTests.GetSolution
CodeStyleOption<T>.FromXElement
Extensions.ToReportDiagnostic
AbstractTriviaFormatter.GetRuleSpacesOrIndentation
AbstractTriviaFormatter.GetRuleSpacesOrIndentation //second switch on a different type in the same method
SerializableBytes.Seek
CsharpSyntaxGenerator.GetParameterModifiers
CsharpSyntaxGenerator.GetTokenKind
```

That is 24 locations.
In all but 3, the `_ =>` arm threw an exception. This means the exhaustiveness checking is gaining the developer nothing since an exception would be thrown even without that arm.

In `FindAllDeclarationsTests.GetSolution` null is returned by the `_ =>` arm. This is a private method that is used 16 times. In every single one of those 16 usages, if null had been returned, a NRE or an ANE would have been thrown later on. In other words, the default behavior of throwing an exception if not all cases are accounted for would have been better than the solution the developer chose to do instead, since it would prevent cascading errors.

`VSTelemetryLogger.CreateAndStartScope` and `VSTelemetryLogger.LogBlockEnd` are more interesting. As loggers they are coded extremely defensively, and throwing an exception for them is useless, since it wont be reported anywhere. Instead on the invalid arm they report the exception rather than throwing it:
```csharp
            try
            {
                // guard us from exception thrown by telemetry
                var kind = kvLogMessage.Kind;
                switch (kind)
                {
                    case LogType.Trace:
                        EndScope<OperationEvent>(functionId, blockId, kvLogMessage, cancellationToken);
                        return;
                    case LogType.UserAction:
                        EndScope<UserTaskEvent>(functionId, blockId, kvLogMessage, cancellationToken);
                        return;
                    default:
                        FatalError.Report(new Exception($"unknown type: {kind}"));
                        break;
                }
            }
            catch
            {
            }
```

In conclusion, in 21 out of 24 cases the strict exhaustiveness checking caused the same code to be emitted as it would have been under the pragmatic exhaustiveness checking. In 1 case it led to worse code being emitted, as it lead to cascading errors, rather than errors at the point where they should have been reported.

In 2 cases it lead the developer to write better code for their use case. But this was in a class that was anyway very defensively written, and not your typical use case.

In all 24 cases there will be no warning if another enum value is added which the switch does not account for, which there would have been under pragmatic exhaustiveness checking in 22 of those cases.

It seems clear to me that overwhelmingly, more bugs will be caught if csharp uses pragmatic exhaustiveness checking rather than strict.

#### C# 8.1

It is very close to C# 8.0 release - possibly too close to make this change. It is not an error to remove a warning later, so this could be changed in C# 8.1 if necessary.

#### InvalidOperation is not a very good exception

It may be pointed out that I've been unfair here. By default the compiler will emit an InvalidOperationException with no information as to what caused the exception. A developer can choose to throw a more useful exception. A good example might be the [InvalidEnumArgumentException](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.invalidenumargumentexception?view=netframework-4.8) which stores the name of the enum, the type of the enum, and the value of the enum.

The compiler might not be able to emit these instead of InvalidOperationException due to the security risk of revealing the names of locals. Also it would have to be a new exception called InvalidEnumSwitchException or something similiar, since the switch might not be on an argument. However if that is the concern, it might be possible for the developer to opt in to which exception should be thrown using attibutes, or some other method. Perhaps an assembly attribute could be overriden by a class attribute could be overriden by a member attribute, etc.

**Comment by spydacarnage** [comment]
I could get behind a solution where the warning doesn't show if all known
enum values are catered for, but if the _ => arm isn't specified, the
compiler adds it with the new InvalidEnumSwitchException being thrown.
With regards to the risk of exposing locals, the exception could just have
the enum value as a property, relying on the developer to use the stack
trace to determine where the actual error occurred (which will, in turn,
determine the name and type of the enum).

**Comment by CyrusNajmabadi** [comment]
I'd prefer the compiler be loose here.  And literally ship with analyzers in-the-box that allowed people to just state that they wanted more strictness.

Specifically i would prefer the following:

1. compiler ships with *no* warnings about exhaustiveness for switch expressions.
2. we have a roslyn analyzer (which should ship in-box in VS as well) with three options.  "No switch exhaustiveness warnings", "warn if any declared switch value is missing" "warn if any possible switch value is missing".

users then opt into the level of checking they want.

**Comment by CyrusNajmabadi** [comment]
Tagging @jcouv .

**Comment by YairHalberstadt** [comment]
A further point is that in about 80% of all cases the _ => arm just threw, even if not all declared enums were accounted for, so I think theres a very strong case to be made that no exhaustiveness checking is a good default.

**Comment by spydacarnage** [comment]
@CyrusNajmabadi - that works.  One of the things I liked about the switch expression was the F#-Match-like exhaustiveness checking, so as long as it is still there in analyzer form, I'd be happy.

**Comment by canton7** [comment]
> we have a roslyn analyzer (which should ship in-box in VS as well) with three options. "No switch exhaustiveness warnings", "warn if any declared switch value is missing" "warn if any possible switch value is missing".

So long as this runs as part of the build process, and not just when the offending file is open. The value here is in being able to change an enum and see all switches which are affected, wherever in the solution they might be.

**Comment by YairHalberstadt** [comment]
Some other cases that ought to be considered:

a) The following warns, since there might be IVTs. But in 99% of cases that warning is unhelpful:
```csharp
using System;
public class P {
    public string M() 
    {
        A a = GetA();
        return a switch {
                B b => "b",
                C c => "c",
                null => "null",
        };
    }

    private A GetA()
    {
        throw new NotImplementedException();
    }
}

internal abstract class A{}
internal class B : A {}
internal class C : A {}
```
b) Flags are likely to be treated differently to other enums, as they are very difficult to treat exhaustively.

c) Someone might want to have a catch all where he throws a custom exception, but still warn if he doesn't explicitly handle all enums.

I think all of this is pointing to the fact that exhaustability checking for switch expressions is not a one size fits all matter. I think it makes most sense to leave it out the language entirely, but ship analysers with roslyn by default, which can be configured to your preferences. They also could pick up on attributes/comments to adjust the settings for a specific switch.

**Comment by CyrusNajmabadi** [comment]
> So long as this runs as part of the build process, and not just when the offending file is open.

Yes.  That's how analyzers work. :)

**Comment by CyrusNajmabadi** [comment]
@gafter @jcouv @MadsTorgersen as i know there is very little time, but i do think this warrants attention for 8.0.

**Comment by HaloFour** [comment]
The impetus for this discussion I believe was #2667 where the OP wanted a feature that would allow them to associate a value of an enum with a result and have the compiler enforce that every value of the enum would be matched, particularly in the case where a new value is added to the enum after the fact.  They had a specific feature in mind based on one from Delphi but the conversation turned to `switch` expressions.

The problem is that with the current exhaustiveness check you have two options:

1.  Leave out the catch-all, in which case the compiler will always warn and that isn't helpful at all.
2. Have the catch-all, in which case the compiler will never warn as the catch-all will handle missed enum values, and that isn't helpful at all.

You could go with 2 and have analyzers do a second pass on exhaustiveness checking, but at that point what's the benefit of having the compiler do the work?  You'd still be required to supply the catch-all to silence the compiler, and odds are the only thing you could do in that case is `throw` anyway.

**Comment by canton7** [comment]
> Yes. That's how analyzers work. :)

It's not how IDE0010 works, so wanted to make sure!

**Comment by CyrusNajmabadi** [comment]
> It's not how IDE0010 works, so wanted to make sure!

Understood :)

**Comment by gafter** [comment]
> there is very little time, but i do think this warrants attention for 8.0.

C# 8 is done.  Only bug fixes for serious issues like crashes or bad code generation can possibly be included.

**Comment by RemmondSoftTechInc** [comment]
I'm the original OP of #2667. I wasn't aware of the exhaustiveness check of the switch statement, which is interesting, but it's not what I was looking for in #2667.  I'm not sure how I feel about my original idea of using enums as array indexes evolving into enhancing the switch statement. I'm concerned the switch behavior will be changed, risking backward compatibility, and I'm no further ahead in my original requirement.  I was looking for a clean/simple way to define lookup arrays indexed on an enum (simple enums starting at 0) that would get the compiler to report an error if there's arrays that are missing values (similar to what it does now if you have a set sized array, see original post for more info).

**Comment by HaloFour** [comment]
@RemmondSoftTechInc 

> I'm not sure how I feel about my original idea of using enums as array indexes evolving into enhancing the switch statement.

The `switch` expression was already being added to the language, along with exhaustiveness checking.  The question was whether or not that exhaustiveness checking as it applied to matching enums would satisfy the desire to know if you've matched all possible values of that enum.  Unfortunately it sounds like the answer is "no", at least for the C# 8.0 timeframe.

Your proposal would be evaluated separately from `switch` expressions or pattern matching in general.

**Comment by CyrusNajmabadi** [comment]
> C# 8 is done. Only bug fixes for serious issues like crashes or bad code generation can possibly be included.

@gafter Thanks Neal.  I personally think this makes sense.  It also feels better to start with a warning that can be potentially relaxed in the future.  That's far safer than not having the warning but then warning to add it.

**Comment by Eirenarch** [comment]
Any update on this? Is changing the behavior considered for a future C# version. I think the proposed behavior is better than the current behavior in every way and what is more it seems to me that changing the behavior here would not introduce a breaking change

**Comment by gafter** [comment]
We have a separate proposal for being able to declare an enum type "closed".  See https://github.com/dotnet/csharplang/issues/3179.

**Comment by Eirenarch** [comment]
I fail to see the value in this in the context of switch exhaustiveness checks. If we agree that this exhaustiveness behavior is desirable why not enable it for all enums?

**Comment by HaloFour** [comment]
@Eirenarch 

All enums are open today.  There's nothing in the language or runtime which prevents you from passing any arbitrary integral value as a value of an enum, and it's not that uncommon for flags enums to be combined in ways that aren't explicitly defined as a value in that enum.  Adding new values isn't necessarily considered a breaking change, which it would have to be in order for the enum to be considered closed.

**Comment by YairHalberstadt** [comment]
@gafter 

Nonetheless, regular enums are not going to go away, and I think this issue is important to improve the experience when switching on them.

**Comment by Eirenarch** [comment]
I understand that. I don't understand the value of exhaustiveness checks as currently implemented for enums. The thing is it is never better to have the exhaustiveness checks work this way so why not change it? Yes, closed enums will allow me to sidestep the issue in my code and maybe a bunch of libraries will adopt them but who needs the behavior of exhaustiveness checks for open enums to begin with? Why keep the worse behavior even if a work-around is provided?

**Comment by HaloFour** [comment]
@Eirenarch 

That value is in forcing the developer to deal with the unexpected vs. just having it throw an exception at runtime.

**Comment by CyrusNajmabadi** [comment]
> The thing is it is never better to have the exhaustiveness checks work this way so why not change it?

I disagree.  There's a *lot* of value here because it makes me think about what can happen about the very real case of a new value flowing through in the future that i may not have thought about at this point in time.

**Comment by Eirenarch** [comment]
But the way to deal with the unexpected is in practice throwing exception and the much more valuable warning for lack of exhaustiveness when a new value is added is lost. It will continue to be lost for open enums.

**Comment by CyrusNajmabadi** [comment]
> But the way to deal with the unexpected is in practice throwing exception

What exception should be thrown?  How would the language know the appropriate exception that i want or need to throw in my codebase for these scenarios?

> and the much more valuable warning for lack of exhaustiveness when a new value is added is lost.

That's easy to solve.  Just write a simple analyzer that checks if there's an explicit enum value you're not writing a case for.  Have it warn then.  Now you have that behavior.

**Comment by Eirenarch** [comment]
The exception that should be thrown is what is thrown right now if I leave it with a warning. Also I don't have the option to not write a default case right now. If we turn your argument around an analyzer should be written for the full exhaustiveness case that forces the developer to always write a default clause because a default clause could always be added but I can't write an analyzer to delete a compiler warning.

**Comment by YairHalberstadt** [comment]
@CyrusNajmabadi 

I took the trouble to gather some data on which is more useful when I created this issue. I think it's clear that it's overwhelmingly more useful not to warn.

**Comment by alrz** [comment]
>That's easy to solve. Just write a simple analyzer that checks if there's an explicit enum value you're not writing a case for. 

I don't think that's too easy though. The compiler can handle nested patterns out of the box - in case enum values appeared in recursive patterns, for example. For an analyzer you have to reconstruct a whole dag for only for this purpose. AFAIK roslyn does not expose such APIs.

As an aside: Rust language that had exhaustive enums from the begining, had to introduce open enums for when you don't want to break client code by adding enum members. So both are useful at diffrenet circumstances. We're just in the reversed situation where we have open enums already.

**Comment by CyrusNajmabadi** [comment]
> We're just in the reversed situation where we have open enums already.

Sure.  I would be fine with a lang feature stating that an enum was closed.  Sounds like a different feature though.

**Comment by CyrusNajmabadi** [comment]
>If we turn your argument around 

But i want *by default* the language to be safer.  So teh direction of hte argument here is very relevant.  by default it leans toward safety.  If you don't want that, add the default check, and then add an analyzer to let you know if you're missing values if that is an important scenario for you.

**Comment by YairHalberstadt** [comment]
> But i want by default the language to be safer.

I would argue that when we actually look at how enums are used in switch expressions, the default ends up being less safe, since developers accidentally leave out enum cases, and don't realise because they have a default. The default just throws, which is what would happen automatically anyway.

**Comment by CyrusNajmabadi** [comment]
> The default just throws, which is what would happen automatically anyway.

I don't agree with that.   Throwing some random internal exception is very different from writing domain specific error handling appropriate for that component.

**Comment by YairHalberstadt** [comment]
> Throwing some random internal exception is very different from writing domain specific error handling appropriate for that component.

In practice that's not what happens. Take a look at Roslyn. They mostly just throw a generic InvalidoperationException.

**Comment by CyrusNajmabadi** [comment]
>  since developers accidentally leave out enum cases

How do you distinguish: "they accidentally left this out", and "they purposefully left it out"?

**Comment by Eirenarch** [comment]
That's like Java's checked exceptions which are supposed to help devs handle some cases but in practice result in empty catches and wrappers for runtime exceptions. In practice these domain exceptions don't exist (and in my experience don't have much of a reason to exist)

**Comment by CyrusNajmabadi** [comment]
> In practice these domain exceptions don't exist (and in my experience don't have much of a reason to exist)

I don't think we can make such a blanket statement for the entire ecosystem which has been built up around this concept for the last 20 years.  

As i said above, i would be amenable for a way for an author to state that an enum was closed.  But i'm not ok with going and making that the default for something which has behaved the exact opposite way this entire time.

**Comment by YairHalberstadt** [comment]
@CyrusNajmabadi 

This issue is not about making enums closed. It's about tooling reflecting the fact that in practice, in the vast majority of cases, developers treat enums as closed, and do not try to handle invalid enum values gracefully.

**Comment by Eirenarch** [comment]
@CyrusNajmabadi I don't expect the ecosystem to provide a lot of examples where the exception is something specific rather than something generic like InvalidOperationException.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi I don't expect the ecosystem to provide a lot of examples where the exception is something specific rather than something generic like InvalidOperationException.

If we have customers using and expecting a different exeption than that, then this would impact them.

As i said above, i don't see how we could change this safely for the entire ecosystem.  If we wanted this, it would need to be an explicit opt-in for enums that are willing to state that they are closed and should have this known behavior. Otherwise we have no idea of knowing if an enum is actually closed, or is just a fancy-named open integer alias.

**Comment by Eirenarch** [comment]
What I am saying is that these domain specific exceptions do not exist. People just throw in some common exception (I put Exception). This could safely be changed because code which has a default handler with a specific exception will be perfectly valid and continue to throw the exception that is written in the code.

**Comment by CyrusNajmabadi** [comment]
> This could safely be changed 

No it couldn't.  It could be that those users expect to throw a specific exception and have become accustomed to having their switches warn about non-exhaustion which they then add a particular throw for.  They're losing that behavior which can be very unexpected and may lead to bugs.

>What I am saying is that these domain specific exceptions do not exist.

Please provide definitive evidence that that's the case. :)

**Comment by Eirenarch** [comment]
Nobody is used to this yet and also they can probably remember to do it without a warning. Again the custom exception case is extremely rare so usability for 99% of the cases should not be sacrificed for usability in 1% of the cases.

As for evidence we have example in the original post. The real world logger code reports Exception (i.e. no specific domain exception although it does not throw). I checked several of the mentioned Roslyn types. All of them use this helper - https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/InternalUtilities/ExceptionUtilities.cs - which throws InvalidOperationException

**Comment by theunrepentantgeek** [comment]
> What I am saying is that these domain specific exceptions do not exist.

Yep, they do.

Regrettably it's in code that belongs to a former employer (so I can't copy/paste and demonstrate), but I would _routinely_ define custom exception types for library code. Giving useful diagnostics in call stacks is utterly vital.

> People just throw in some common exception (I put Exception). 

And that would be a hard reject on any - every - pull request I reviewed. There are a few justifiable cases for catching `Exception`, but I can't recall a good reason to throw it (and I've heard dozens).

**Comment by Eirenarch** [comment]
@theunrepentantgeek you have a lot of domain specific exceptions for default cases in switches which fully cover the declared enum values? I know that domain specific exceptions exist, I just don't think they are used for guarding against unknown values in enum-based switches

**Comment by theunrepentantgeek** [comment]
Define "a lot". 😉 

If I define an enumeration where being exhaustive in my switch statements is necessary, I want to make damn sure that there are good diagnostics if/when a case is missed. It takes just a few minutes to create a custom exception type, and it protects me from my own mistakes, as well as protecting future developers who don't have the context that I do.

I don't do it for every enumeration. But it is common for me to do so.

> What I am saying is that these domain specific exceptions _**do not exist.**_
(emphasis added)

This is incorrect. They do.

**Comment by Eirenarch** [comment]
Do you have specific handling for them or they are simply diagnostics tools?

**Comment by mqudsi** [comment]
I'm not sure if this is the point that was being made in one of the earlier comments or not (and I apologize if so) but I think requiring a `_ =>` match is actually harmful. If an enum has two members today and they're both handled, warning-free code looks like this today:

```csharp

public enum Foo {
    Foo,
    Bar,
}

public int Bar(Foo foo) {
    return foo switch {
        Foo.Foo => 0,
        Foo.Bar => 1,
        _ => throw new InvalidOperationException(),
    }
}
```

Tomorrow a new case is added to the enum, and the code ends up looking like this:

```csharp

public enum Foo {
    Foo,
    Bar,
    Baz,
}

public int Bar(Foo foo) {
    return foo switch {
        Foo.Foo => 0,
        Foo.Bar => 1,
        _ => throw new InvalidOperationException(),
    }
}
```

The new case isn't handled, but no warning is emitted because the catch-all was already present.

I think the correct approach is for the compiler to emit a warning when no `_ => ` case exists and not all explicitly named enum values are present, and to then shim the additional `_ => throw ....` case (as it does). This isn't about ergonomics, reducing boilerplate, tamping down on annoying warnings, or anything else - it's to prevent bugs from creeping in.

**Comment by CyrusNajmabadi** [comment]
> The new case isn't handled, but no warning is emitted because the catch-all was already present.

Why is that a problem?  Code that handles enums should handle that they are open ended.  Your code coudl always run against the newer enum without it being recompiled.  So you have to be resilient to this no matter what.

**Comment by Eirenarch** [comment]
It is a problem because the compiler doesn't help me find all the places where I should change my code. The code could not run against the new value because in the vast majority of cases it throws an exception.

**Comment by yaakov-h** [comment]
That sounds like a good use-case for an analyzer to insert warnings.

**Comment by Eirenarch** [comment]
The problem is we'll also need an analyzer to remove warnings

**Comment by yaakov-h** [comment]
How so?

**Comment by Eirenarch** [comment]
Because I don't want to have to write a default case to suppress a warning. Now technically I can suppress the warning in the config but then I am suppressing it for values like ints where I want it. In addition it feels super strange that the case for 99% of users would be to suppress a warning and then add an analyzer.

**Comment by mqudsi** [comment]
I‘ll disagree with @Eirenarch here, my reasoning is simply that there is much more merit in Roslyn warning about an explicitly extant enum that isn’t covered in a switch than there is in it always warning it the default branch isn’t covered. I.E. if anything I feel one should “write an analyzer” to force explicit handling of the _ branch rather than the other way around.

Whether the goal is to maximize utility or to maximize safety, I feel the developer’s interests are best served by prioritizing the dual logic/safety error of “explicit enum not covered” over “default behavior not overridden” (because it *is* implicitly covered by the codegen with the throwing behavior).

**Comment by Eirenarch** [comment]
@mqudsi I don't see how you disagree with me. I've been saying the same thing all through this thread.

**Comment by mqudsi** [comment]
Sorry, I meant about wanting to remove warnings. I don’t have a problem writing extra code to please the compiler (in the presence of a theoretical extra analyzer to check for unused enums I would not be sufficiently irked  by the extra branch being required to want to strip that warning out).  But we absolutely agree on what the ideal behavior should look like, I think.

**Comment by Eirenarch** [comment]
The problem is that not only is this code bloat (a reminder that the compiler will indeed add the exception case for you) but that if you really want to handle some cases via the default clause you can't because it will result in a warning

**Comment by gafter** [comment]
Please see https://github.com/dotnet/csharplang/issues/3179

**Comment by stephan-tolksdorf** [comment]
> How do you distinguish: "they accidentally left this out", and "they purposefully left it out"?

Generate two different warnings for the case where a switch doesn't handle a named enum value and for the case where the switch doesn't handle a value for which no enum constant exists. Then I can selectively suppress the latter warning and still get a compiler warning when somebody forgets updating a switch after adding a new enum constant.

**Comment by CyrusNajmabadi** [comment]
This feels like something an analyzer could really simply do.  It would even be a good example used to teach people how to write analyzers (esp. IOp analyzers).

**Comment by gafter** [comment]
Once https://github.com/dotnet/roslyn/pull/44702 is integrated, there will be different warnings when a named enum value is not handled vs when an unnamed enum value is not handled. In the former case the compiler will tell you the name of an enum constant that is not handled. In the latter case it will say something like `For example '(EnumType)23' is not handled`.

**Comment by binki** [comment]
@gafter
> Once [dotnet/roslyn#44702](https://github.com/dotnet/roslyn/pull/44702) is integrated, there will be different warnings when a named enum value is not handled vs when an unnamed enum value is not handled. In the former case the compiler will tell you the name of an enum constant that is not handled. In the latter case it will say something like `For example '(EnumType)23' is not handled`.

I was hoping that by “different warnings” you meant different diagnostic codes. Then I could blacklist the the named warning with `WarningsAsErrors` and then silence the other one (since I would rather catch bad data at the edge/load/parse-time and am fine with a runtime explosion if it somehow gets in because that should be impossible). However, both warnings use the same code of CS8509, so I can’t distinguish them without parsing post-translation/human-readable text…

    C:\Users\ohnob\repos\Main 33\Code\«redacted».cs(1207,25,1207,31): error CS8509: The switch expression does not handle all possible values of its input type (it is not exhaustive). For example, the pattern '(System.Data.DbType)24' is not covered.

If we had different warnings for the undefined enum case and the defined case, that would solve this for me.

**Reply by binki** [reply]
Looks like this happened in https://github.com/dotnet/roslyn/issues/47066 . Yay, thanks @gafter and @dextercd!

**Comment by jonkuhn** [comment]
I agree with @binki I would really like to be able to have a way to write a pattern matching switch statement such that the compiler would error if a new value was added to the enum that was not handled.  If it were possible to write the switch statement without the default case and then ignore one warning (for not handling every possible runtime value), but configure another warning as an error (for not handling a named enum value) that would solve this use case.

**Comment by nmoreaud** [comment]
I think the integer value warning should be off by default, this is cryptic: `For example '(EnumType)23' is not handled`.

**Reply by binki** [reply]
That error is pretty clear to me. You do understand that the following [is valid code](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAgrgOwD4AEBMAjALABQZAqlHALICeAopumABT1MsB0AggJQBuStQ7M27RmL4AWQWTK5pAAiq1JLCZzBKw6sHyUBvMktNLchAJysAJACJDurQF8AvIdYBLTDD5OxznZypM7ypHBiSqIsRiZmPEhxpgBCiSFkQA), right? How else would the error be written to be clearer?

```csharp
using System;

UseMyEnum(MyEnum.A);
UseMyEnum((MyEnum)4);

void UseMyEnum(MyEnum myEnum) {
    Console.WriteLine($"{myEnum}={(int)myEnum}");
}

enum MyEnum {
    A,
    B,
}
```

Output:
```
A=0
4=4
```

**Reply by nmoreaud** [reply]
Sure this is valid, but it is not the main use case.
If I modify my code to get rid of the warning, I will forget to add case entries for new enum values

**Reply by HaloFour** [reply]
The compiler intentionally emits two different warnings, `CS8524` if you don't cover "invalid" enum values and `CS8509` if you didn't cover all of the values declared in the enum.  You can tell the compiler to ignore one and not the other.

**Comment by jnm2** [comment]
InvalidEnumArgumentException should only be thrown if the thing being switched on is a method argument.

**Reply by binki** [reply]
What’re you replying to? ;-)

**Reply by jnm2** [reply]
I'm replying to the post.

> **InvalidOperation is not a very good exception**
> 
> It may be pointed out that I've been unfair here. By default the compiler will emit an InvalidOperationException with no information as to what caused the exception. A developer can choose to throw a more useful exception. A good example might be the InvalidEnumArgumentException which stores the name of the enum, the type of the enum, and the value of the enum.

**Reply by jnm2** [reply]
Oh, I missed this sentence which clears it up:

> Also it would have to be a new exception called InvalidEnumSwitchException or something similiar, since the switch might not be on an argument.

**Comment by fernandozago** [comment]
## Improve enum exhaustiveness checking

I often need to distinguish between:

1. invalid/unnamed enum values, such as `(MyTypes)10`
2. valid named enum values that I forgot to handle after the enum changes

Today I end up doing this:

```csharp
MyTypes t = (MyTypes)10;

#pragma warning disable CS8524 // Switch expression does not handle unnamed enum values.
string val = !Enum.IsDefined(t) ? "OutOfRange" : t switch
{
    MyTypes.A => "InRange",
    MyTypes.B => "InRange",
    MyTypes.C => "InRange",
    MyTypes.E => "InRange",
    MyTypes.F => "InRange",
};
#pragma warning restore CS8524

Console.WriteLine(val);

public enum MyTypes
{
    A,
    B,
    C,
    E,
    F,
    // Z, // Uncommenting this correctly produces CS8509.
}
```

This gives me the behavior I want:

* `(MyTypes)10` is handled as `"OutOfRange"`.
* Adding a new named enum member, for example `Z`, still produces `CS8509`.

The problem is that I need to suppress `CS8524`, even though unnamed enum values were already handled by `Enum.IsDefined(t)`.

It would be useful to express this directly in the switch expression.

One possible direction, related to https://github.com/dotnet/roslyn/issues/45157, would be to distinguish named enum fallback from unnamed enum fallback:

```csharp
string val = t switch
{
    MyTypes.A => "InRange",
    MyTypes.B => "InRange",
    MyTypes.C => "InRange",
    MyTypes.E => "InRange",
    MyTypes.F => "InRange",

    // Remaining declared enum members only.
    // If absent, adding a new enum member should still produce CS8509.
    _named => "InRange_NotMapped",

    // Invalid/unnamed enum values only, such as (MyTypes)10.
    // This should satisfy CS8524 without hiding CS8509.
    _unnamed => "OutOfRange"
};
```

Another possible direction would be allowing the compiler to understand a guarded arm like this as handling unnamed enum values only:

```csharp
string val = t switch
{
    MyTypes.A => "InRange",
    MyTypes.B => "InRange",
    MyTypes.C => "InRange",
    MyTypes.E => "InRange",
    MyTypes.F => "InRange",

    // Handles undefined enum values, but should not hide new named values.
    var value when !Enum.IsDefined(value) => "OutOfRange"
};
```

In both examples, the intent is the same:

> Undefined enum values should be handled, but newly added named enum members should still produce `CS8509` unless explicitly handled.

Today, using a normal discard arm does not work for this:

```csharp
string val = t switch
{
    MyTypes.A => "InRange",
    MyTypes.B => "InRange",
    MyTypes.C => "InRange",
    MyTypes.E => "InRange",
    MyTypes.F => "InRange",
    _ => "OutOfRange"
};
```

The `_` arm handles unnamed enum values, but it also hides future named enum members. If `MyTypes.Z` is added later, the compiler can no longer report `CS8509`.

Would it be reasonable to improve enum exhaustiveness support so developers can express “exhaustive over named enum values only” without suppressing `CS8524`, and without using a discard arm that hides future named enum members?

**Comment by Flutterish** [comment]
Zig handles this by using a `_ =>` to exhaustively check an enum, or `else =>` to non-exhaustively check an enum. Hovewer that specific syntax makes sense because you also use `_` as an enum "memeber" to declare it as a non-exhaustive enum. Regular zig enums are exhaustive and switches normally dont need an `else` branch because if you somehow make a value outside the declared ones, thats UB (as it should be).
